doc/cascadia/Json-Utility-API.md

```diff
@@ -0,0 +1,163 @@
+
+auto three = JsonUtils::GetValueForKey<std::string>(json, "thirdKey");
+// three is populated or zero-initialized
+
+auto four = JsonUtils::GetValueForKey<std::optional<std::string>>(json, "fourthKey");
+// four is populated or nullopt
+```
+
+## Converting User-Defined Types
+
+All conversions are done using specializations of `JsonUtils::ConversionTrait<T>`.
+To implement a converter for a user-defined type, you must implement a specialization of `JsonUtils::ConversionTrait<T>`.
+
+Every specialization over `T` must implement `static T FromJson(const Json::Value&)` and
+`static bool CanConvert(const Json::Value&)`.
+
+```c++
+template<>
+struct ConversionTrait<MyCustomType>
+{
+    // This trait converts a string of the format "[0-9]" to a value of type MyCustomType.
+
+    static MyCustomType FromJson(const Json::Value& json)
+    {
+        return MyCustomType{ json.asString()[0] - '0' };
+    }
+
+    static bool CanConvert(const Json::Value& json)
+    {
+        return json.isString();
+    }
+};
+```
+
+For your "convenience" (;P), if you need to provide name-value enum mapping,
+there's the `EnumMapper<>` base template. It is somewhat verbose.
+
+```c++
+template<>
+struct JsonUtils::ConversionTrait<CursorStyle> : public JsonUtils::EnumMapper<CursorStyle, JsonUtils::ConversionTrait<CursorStyle>>
+{
+    // Unfortunately, you need to repeat the enum type three whole times ^
+
+    // pair_type is provided by EnumMapper to make your life easier.
+    // Unfortunately, you need to provide  v- there  a count of the values in the enum.
+    static constexpr std::array<pair_type, 5> mappings = {
+        pair_type{ "bar", CursorStyle::Bar }, // DEFAULT
+        pair_type{ "vintage", CursorStyle::Vintage },
+        pair_type{ "underscore", CursorStyle::Underscore },
+        pair_type{ "filledBox", CursorStyle::FilledBox },
+        pair_type{ "emptyBox", CursorStyle::EmptyBox }
+    };
+};
+```
+
+### Advanced Use
+
+`GetValue` and `GetValueForKey` can be passed, as their final arguments, any value whose type implements the same
+interface as `ConversionTrait<T>`--that is, `FromJson(const Json::Value&)` and `CanConvert(const Json::Value&)`.
+
+This allows for one-off conversions without a specialization of `ConversionTrait` or even stateful converters.
+
+#### Stateful Converter Sample
+
+```c++
+struct MultiplyingConverter {
+    int BaseValue;
+
+    bool CanConvert(const Json::Value&) { return true; }
+
+    int FromJson(const Json::Value& value)
+    {
+        return value.asInt() * BaseValue;
+    }
+};
+
+...
+
+Json::Value jv = /* a json value containing 66 */;
+MultiplyingConverter conv{10};
+
+auto v = JsonUtils::GetValue<int>(jv);
+// v is equal to 660.
+```
+
+## Behavior Chart
+
+### GetValue(T&) (type-deducing)
+
+-|json type invalid|json null|valid
+-|-|-|-
+`T`|❌ exception|🔵 unchanged|✔ converted
+`std::optional<T>`|❌ exception|🟨 `nullopt`|✔ converted
+
+### GetValue&lt;T&gt;() (returning)

```

Comment:
So I'm not totally sure _when_ we'd use the _returning_ variations on these, but they make sense to me logically